src/rust/engine/serverset/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+// Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+// Enable all clippy lints except for many of the pedantic ones. It's a shame this needs to be copied and pasted across crates, but there doesn't appear to be a way to include inner attributes from a common source.
+#![cfg_attr(
+  feature = "cargo-clippy",
+  deny(
+    clippy,
+    default_trait_access,
+    expl_impl_clone_on_copy,
+    if_not_else,
+    needless_continue,
+    single_match_else,
+    unseparated_literal_suffix,
+    used_underscore_binding
+  )
+)]
+// It is often more clear to show that nothing is being moved.
+#![cfg_attr(feature = "cargo-clippy", allow(match_ref_pats))]
+// Subjective style.
+#![cfg_attr(
+  feature = "cargo-clippy",
+  allow(len_without_is_empty, redundant_field_names)
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![cfg_attr(
+  feature = "cargo-clippy",
+  allow(new_without_default, new_without_default_derive)
+)]
+
+extern crate num_rational;
+extern crate parking_lot;
+
+use parking_lot::Mutex;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+///
+/// A collection of resources which are observed to be healthy or unhealthy.
+/// Getting the next resource skips any which are marked unhealthy, and will re-try unhealthy
+/// resources at an exponentially backed off interval. Unhealthy resources marked healthy will ease
+/// back into rotation with exponential ease-in.
+///
+pub struct Serverset<T> {
+  inner: Arc<Inner<T>>,
+}
+
+impl<T> Clone for Serverset<T> {
+  fn clone(&self) -> Self {
+    Serverset {
+      inner: self.inner.clone(),
+    }
+  }
+}
+
+struct Inner<T> {
+  servers: Vec<Backend<T>>,
+
+  // Only visible for testing
+  pub(crate) next: AtomicUsize,
+
+  failure_initial_lame: Duration,
+
+  failure_backoff_ratio: num_rational::Ratio<u32>,
+
+  failure_max_lame: Duration,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Health {
+  Healthy,
+  Unhealthy,
+}
+
+#[derive(Debug)]
+struct UnhealthyInfo {
+  unhealthy_since: Instant,
+  next_attempt_after: Duration,
+}
+
+#[derive(Debug)]
+struct Backend<T> {
+  server: T,
+  unhealthy_info: Arc<Mutex<Option<UnhealthyInfo>>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct BackoffConfig {
```

Comment:
Could/should we use https://docs.rs/tokio-retry/0.2.0/tokio_retry/ here?

Reply:
There's a comment about avoiding futures here because fork safety; when we're out of our fork world, I will gladly rewrite this to use `tokio_retry` :)